tests/sharp_interfaces.cc

```diff
@@ -0,0 +1,642 @@
+                              surface_fe,
+                              surface_quad,
+                              dof_handler.get_triangulation(),
+                              mapping);
+
+  // step 2) loop over all cells and evaluate curvature and normal in the cell-local
+  // quadrature points
+  //   and test with all test functions of the cell
+
+  AffineConstraints<double> constraints; // TODO: use the right ones
+
+  FEPointEvaluation<1, dim> phi_curvature(mapping, dof_handler.get_fe());
+
+  FESystem<dim>               fe_dim(dof_handler.get_fe(), dim);
+  FEPointEvaluation<dim, dim> phi_normal_force(mapping, fe_dim);
+
+  std::vector<double>                  buffer;
+  std::vector<double>                  buffer_dim;
+  std::vector<types::global_dof_index> local_dof_indices;
+
+  for (unsigned int i = 0; i < cells.size(); ++i)
+    {
+      typename DoFHandler<dim>::active_cell_iterator cell = {
+        &dof_handler.get_triangulation(), cells[i].first, cells[i].second, &dof_handler};
+
+      const unsigned int n_dofs_per_component = cell->get_fe().n_dofs_per_cell();
+
+      local_dof_indices.resize(n_dofs_per_component);
+      buffer.resize(n_dofs_per_component);
+      buffer_dim.resize(n_dofs_per_component * dim);
+
+      cell->get_dof_indices(local_dof_indices);
+
+      const unsigned int n_points = ptrs[i + 1] - ptrs[i];
+
+      const ArrayView<const Point<dim>> unit_points(points.data() + ptrs[i], n_points);
+      const ArrayView<const double>     JxW(weights.data() + ptrs[i], n_points);
+
+      // gather curvature
+      constraints.get_dof_values(curvature_solution,
+                                 local_dof_indices.begin(),
+                                 buffer.begin(),
+                                 buffer.end());
+
+      // evaluate curvature
+      phi_curvature.evaluate(cell,
+                             unit_points,
+                             make_array_view(buffer),
+                             EvaluationFlags::values);
+
+      // gather normal
+      for (int i = 0; i < dim; ++i)
+        {
+          constraints.get_dof_values(normal_vector_field.block(i),
+                                     local_dof_indices.begin(),
+                                     buffer.begin(),
+                                     buffer.end());
+          for (unsigned int c = 0; c < n_dofs_per_component; ++c)
+            buffer_dim[c * dim + i] = buffer[c];
+        }
+
+      // evaluate normal
+      phi_normal_force.evaluate(cell, unit_points, buffer_dim, EvaluationFlags::values);
+
+      // quadrature loop
+      for (unsigned int q = 0; q < n_points; ++q)
+        {
+          const auto normal =
+            phi_normal_force.get_value(q) / phi_normal_force.get_value(q).norm();
+          phi_normal_force.submit_value(normal * phi_curvature.get_value(q) * JxW[q], q);
+        }
+
+      // integrate force
+      phi_normal_force.integrate(cell, unit_points, buffer_dim, EvaluationFlags::values);
```

Comment:
I have integrate the new changes from https://github.com/dealii/dealii/pull/11630 and https://github.com/dealii/dealii/pull/11623 and now we can write on arbitrary surfaces code that is confusingly similar to code we would write in the context of `MatrixFree`/`FEFaceEvaluation`.

One can argue if the time spent to accomplish this was worth it but I think it is cool! I will think about how to make this even more user-friendly (`JxW`) and how we can integrate `FEPointEvaluation` into the `MatrixFree` ecosystem/workflow.

FYI @kronbichler @mschreter